rust/src/writer/utils.rs

```diff
@@ -12,31 +12,59 @@ use arrow::{
 use parquet::file::writer::InMemoryWriteableCursor;
 use serde_json::Value;
 use std::collections::HashMap;
+use std::fmt::Display;
 use std::io::Write;
 use std::sync::Arc;
 use uuid::Uuid;
 
 const NULL_PARTITION_VALUE_DATA_PATH: &str = "__HIVE_DEFAULT_PARTITION__";
 
-pub(crate) fn get_partition_key(
-    partition_columns: &[String],
-    partition_values: &HashMap<String, Option<String>>,
-) -> Result<String, DeltaWriterError> {
-    let mut path_parts = vec![];
-    for k in partition_columns.iter() {
-        let partition_value = partition_values
-            .get(k)
-            .ok_or_else(|| DeltaWriterError::MissingPartitionColumn(k.to_string()))?;
-
-        let partition_value = partition_value
-            .as_deref()
-            .unwrap_or(NULL_PARTITION_VALUE_DATA_PATH);
-        let part = format!("{}={}", k, partition_value);
-
-        path_parts.push(part);
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub(crate) struct PartitionPath {
+    path: String,
+}
+
+impl PartitionPath {
+    pub fn from_hashmap(
+        partition_columns: &[String],
+        partition_values: &HashMap<String, Option<String>>,
+    ) -> Result<Self, DeltaWriterError> {
+        let mut path_parts = vec![];
+        for k in partition_columns.iter() {
+            let partition_value = partition_values
+                .get(k)
+                .ok_or_else(|| DeltaWriterError::MissingPartitionColumn(k.to_string()))?;
+
+            let partition_value = partition_value
+                .as_deref()
+                .unwrap_or(NULL_PARTITION_VALUE_DATA_PATH);
```

Comment:
Just realized that this crate traditionally uses the constant, but the protocol specifies a different behaviour when it comes to nulls. I opened a ticket to track this question #619 - since it was also not introduced here...

https://github.com/delta-io/delta/blob/master/PROTOCOL.md#partition-value-serialization